x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationClusterService.java

```diff
@@ -356,64 +357,63 @@ ClusterState addRemoveAllocationNodes(ClusterState currentState) {
         final TrainedModelAllocationMetadata previousState = TrainedModelAllocationMetadata.fromState(currentState);
         final TrainedModelAllocationMetadata.Builder builder = TrainedModelAllocationMetadata.builder(currentState);
         Set<String> shuttingDownNodes = nodesShuttingDown(currentState);
-        Set<String> currentNotShuttingDownNodes = currentState.getNodes()
+        Map<String, DiscoveryNode> currentNotShuttingDownNodes = currentState.getNodes()
             .getAllNodes()
             .stream()
-            .map(DiscoveryNode::getId)
-            .filter(id -> shuttingDownNodes.contains(id) == false)
-            .collect(Collectors.toSet());
-        // TODO: make more efficient, right now this is O(nm) where n = sizeof(models) and m = sizeof(nodes)
-        // It could probably be O(max(n, m))
-        // Add nodes and keep track of currently routed nodes
-        // Should we indicate a partial allocation somehow if some nodes don't have space?
-        for (Map.Entry<String, TrainedModelAllocation> modelAllocationEntry : previousState.modelAllocations().entrySet()) {
-            // Don't bother adding/removing nodes if this allocation is stopping
-            if (modelAllocationEntry.getValue().getAllocationState().equals(AllocationState.STOPPING)) {
-                continue;
-            }
-            final String modelId = modelAllocationEntry.getKey();
-            Map<String, String> nodeToReason = new TreeMap<>();
-            for (DiscoveryNode node : currentState.getNodes()) {
-                // Only add the route if the node is NOT shutting down, this would be a weird case of the node
-                // just being added to the cluster and immediately shutting down...
-                if (shuttingDownNodes.contains(node.getId()) == false
-                    && StartTrainedModelDeploymentAction.TaskParams.mayAllocateToNode(node)
-                    && modelAllocationEntry.getValue().isRoutedToNode(node.getId()) == false) {
-                    Optional<String> failure = nodeHasCapacity(currentState, modelAllocationEntry.getValue().getTaskParams(), node);
-                    if (failure.isPresent()) {
-                        nodeToReason.put(node.getName(), failure.get());
-                    } else {
-                        builder.getAllocation(modelId).addNewRoutingEntry(node.getId());
+            .filter(node -> shuttingDownNodes.contains(node.getId()) == false)
+            .collect(Collectors.toMap(DiscoveryNode::getId, Function.identity()));
+        // TODO: make more efficient, we iterate every entry, sorting by nodes routed (fewest to most)
+        previousState.modelAllocations()
+            .entrySet()
+            .stream()
+            .filter(entry -> entry.getValue().getAllocationState().equals(AllocationState.STOPPING) == false)
+            .sorted(Comparator.comparing(e -> e.getValue().getNodeRoutingTable().size()))
```

Comment:
This is obviously not optimal, but I am not 100% sure how else to do it.

One issue here (besides performance), is that we always use the counts based on the original cluster state. Meaning, as models are allocated, our iteration order isn't updated in the same pass. 

This is probably acceptable. 

Another option is that we maybe iterate on nodes in the outer loop? This way all allocations attempt the same node and then we move on? The only downside here is that the order of inner iteration may change all the time.

Another option is attempting this same thing with a priority queue that updates on each iteration, popping off all the least allocations and rebuilding each time, but this seems pretty expensive to me as the whole heap would be rebuit on every pass on a new node...